cmake/Modules/ProtocGenUpb.cmake

```diff
@@ -0,0 +1,389 @@
+  endif()
+else()
+    message(STATUS "Library upb executable protoc-gen-upb found: ${UPB_GENERATOR}")
+endif()
+get_filename_component(UPB_GENERATOR_NAME ${UPB_GENERATOR} NAME_WE)
+get_filename_component(UPB_GENERATOR_DIR ${UPB_GENERATOR} DIRECTORY)
+
+# Find the protoc-gen-upbdefs Executable
+if(NOT UPBDEFS_GENERATOR)
+  find_program(
+    UPBDEFS_GENERATOR
+    NAMES protoc-gen-upbdefs
+    DOC "protoc-gen-upbdefs pluggin to protoc for the generation of μpb (often written 'upb') reflection code"
+    HINTS ${_HINT_PATHS}
+  if(UPBDEFS_GENERATOR)
+    message(STATUS "Library upb executable protoc-gen-upbdefs found: ${UPBDEFS_GENERATOR}")
+  else()
+    message(STATUS "Library upb executable protoc-gen-upbdefs NOT found!")
+  endif()
+else()
+    message(STATUS "Library upb executable protoc-gen-upbupb found: ${UPBDEFS_GENERATOR}")
+  )
+endif()
+get_filename_component(UPBDEFS_GENERATOR_NAME ${UPBDEFS_GENERATOR} NAME_WE)
+get_filename_component(UPBDEFS_GENERATOR_DIR ${UPBDEFS_GENERATOR} DIRECTORY)
+unset(_HINT_PATHS)
+
+# ====================================================================
+#
+# UPB_GENERATE_C (public function)
+# UPB_GENERATE_C(SRCS HDRS [RELPATH <root-path-of-proto-files>] <proto-files>...)
+#   SRCS = Variable to define with autogenerated source files
+#   HDRS = Variable to define with autogenerated header files
+#   RELPATH (optional): Use if you want to use relative paths in your import statements
+#     The argument to RELPATH should be the directory that all the
+#     imports will be relative to. When RELPATH is not specified then all proto
+#     files can be imported without a path.
+#
+# ====================================================================
+# Example:
+#
+# find_package( upb REQUIRED )
+# upb_generate_c(PROTO_SRCS PROTO_HDRS foo.proto)
+#
+# add_executable(bar)
+# target_sources(bar PRIVATE bar.c ${PROTO_SRCS} ${PROTO_HDRS})
+# target_include_directories(bar PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
+# target_link_libraries(bar PUBLIC upb::upb)
+#
+#
+# Example with RELPATH:
+# Assume we have a layout like:
+#  .../CMakeLists.txt
+#  .../bar.cc
+#  .../proto/
+#  .../proto/foo.proto (Which contains: import "sub/bar.proto"; )
+#  .../proto/sub/bar.proto
+#
+# Everything would be the same as the previous example, but the call to
+# upb_generate_c would change to:
+#
+# upb_generate_c(PROTO_SRCS PROTO_HDRS RELPATH proto proto/foo.proto
+# proto/sub/bar.proto)
+#
+# ====================================================================
+
+function(UPB_GENERATE_C SRCS HDRS)
+  # cmake-format: off
+  cmake_parse_arguments(UPB_GENERATE_C "" "RELPATH" "" ${ARGN})
+  if(NOT UPB_GENERATE_C_UNPARSED_ARGUMENTS)
+    message(SEND_ERROR "Error: UPB_GENERATE_C() called without any proto files")
+    return()
+  endif()
+
+  # Construct include list
+  set(_upb_include_set)
+  foreach(_dir ${UPB_INCLUDE_DIRS})
+    list(APPEND _upb_include_set -I${_dir})
+  endforeach()
+
+  if(UPB_GENERATE_X_APPEND_PATH)
```

Comment:
This behavior is harmful; protos should always be referenced by their canonical filename. eg. `google/protobuf/descriptor.proto`, not just `descriptor.proto`.

In particular, only one file with a given name can be loaded at a time. If we have `foo/bar/baz.proto` and automatically add `-Ifoo/bar`, then `baz.proto` will conflict with any other file named `xyz/baz.proto`.

We should not implicitly add `-I` paths.